src/client/ua_client_subscriptions.c

```diff
@@ -1007,6 +1043,70 @@ UA_Client_MonitoredItems_modify_async(UA_Client *client,
     return statusCode;
 }
 
+static void *
+ua_MonitoredItem_findByID(void *data, UA_Client_MonitoredItem *mon) {
+	UA_UInt32 monitorId = *(UA_UInt32*)data;
+	if (monitorId && (mon->monitoredItemId != monitorId)) {
```

Comment:
Here you want to check for equality. Not inequality!
But this is rather inefficient because the ziptree is iterated linearly.
The balanced tree can be searched in log time however.

The methods for that are defined here:
https://github.com/open62541/open62541/blob/31947a93608e6e7d7132f06573ad00c263641b7b/src/client/ua_client_subscriptions.c#L46

Here is an example use:
https://github.com/open62541/open62541/blob/31947a93608e6e7d7132f06573ad00c263641b7b/src/client/ua_client_subscriptions.c#L1067

You can copy / paste the code from the last example and use findMonitoredItemById in that place instead.

Reply:
Ah I see the problem.
We have both MonitoredItemId and ClientHandle.
It is okay to have the linear search for the MonitoredItemId then...